File: Tools/BacktraderReverser/Backtrader.py
import datetime as dt
import numpy as np
import pandas as pd
import backtrader as bt
import math


class LoggerMain:
    def __init__(self):
        self.buf = f'# Main\nDate,Open,High,Low,Close,Equity,Cash'
    def log(self, strat: bt.Strategy):
        broker: bt.BackBroker = strat.broker
        str = f'''
{strat.data.datetime.date()},\
{strat.data.open[0]},\
{strat.data.high[0]},\
{strat.data.low[0]},\
{strat.data.close[0]},\
{broker.get_value()},\
{broker.cash}'''
        #print(str)
        self.buf += str



class LoggerOrder:
    def __init__(self):
        self.buf = f'\n# Orders\nDate,Status,Dir,Size,Type,Price,PriceLimit,ExecSize,ExecPrice,ExecComm'
    def log(self, strat: bt.Strategy, order: bt.Order):
        orderPriceStr = '_' if order.price is None else f'{order.price}'
        orderPriceLimitStr = '_' if order.plimit is None else f'{order.plimit}'
        str = f'''
{strat.data.datetime.date()},\
{order.getstatusname()},\
{order.ordtypename()},\
{order.size},\
{bt.Order.ExecTypes[order.exectype]},\
{orderPriceStr},\
{orderPriceLimitStr},\
{order.executed.size},\
{order.executed.price},\
{order.executed.comm}'''
        #print(str)
        self.buf += str



class LoggerTrade:
    def __init__(self):
        self.buf = f'\n# Trades\nDate,Status,IsLong,Size,Price,PnL,Comm'
    def log(self, strat: bt.Strategy, trade: bt.Trade):
        str = f'''
{strat.data.datetime.date()},\
{bt.Trade.status_names[trade.status]},\
{trade.long},\
{trade.size},\
{trade.price},\
{trade.pnl},\
{trade.pnlcomm}'''
        #print(str)
        self.buf += str



logMain = LoggerMain()
logOrder = LoggerOrder()
logTrade = LoggerTrade()



class SingleOrder(bt.Strategy):
    def next(self):
        logMain.log(self)
        #if self.data.datetime.date() == dt.date(2024, 1, 1):
        #    self.buy(size=10, exectype=bt.Order.Market)
        #if self.data.datetime.date() == dt.date(2024, 1, 3):
        #    self.sell(size=10, exectype=bt.Order.Market)
__orders__
    def notify(self, order: bt.Order):
        logOrder.log(self, order)
    def notify_trade(self, trade: bt.Trade):
        logTrade.log(self, trade)

cerebro = bt.Cerebro(stdstats = True)
cerebro.adddata(bt.feeds.PandasData(dataname=pd.DataFrame(
    {
        #'Open'  : [3, 5, 18, 7, 6, 5, 4],
        #'Close' : [3, 5, 8, 7, 6, 5, 4],
        #'High'  : [3, 5, 8, 7, 6, 5, 4],
        #'Low'   : [3, 5, 8, 7, 6, 5, 4],
__prices__
    },
    index=pd.date_range('__startDate__', periods=__pricesLength__, freq='d', name='Date')
)))
cerebro.addstrategy(SingleOrder)
broker: bt.BackBroker = cerebro.broker
broker.setcash(__brokerCash__)
broker.setcommission(commission=__brokerComm__, leverage=1)
broker.set_coc(__cheatOnClose__)
strats = cerebro.run()

print(logMain.buf)
print(logOrder.buf)
print(logTrade.buf)

File: Tools/BacktraderReverser/Backtrader.cs
using System.Text;
using BacktraderReverser.Structs;
using BacktraderReverser.Utils;

namespace BacktraderReverser;

public sealed class BrokerOpts
{
	public double Cash { get; init; } = 100;
	public double Commission { get; init; } = 0;
	public bool CheatOnClose { get; init; } = false;
}

public static class Backtrader
{
	const string PythonCodeResourceName = "BacktraderReverser.Backtrader.py";

	public static Results Run(
		Prices prices,
		Order[] orders,
		BrokerOpts brokerOpts
	)
	{
		var code = ResourceLoader.Load(PythonCodeResourceName)
			.Replace("__prices__", prices.Fmt())
			.Replace("__startDate__", $"{Consts.StartDate:yyyy-MM-dd}")
			.Replace("__pricesLength__", $"{prices.Length}")
			.Replace("__orders__", orders.Fmt())
			.Replace("__brokerCash__", $"{brokerOpts.Cash}")
			.Replace("__brokerComm__", $"{brokerOpts.Commission}")
			.Replace("__cheatOnClose__", $"{brokerOpts.CheatOnClose}")
			;
		return Python.Run(code).ReadResults();
	}


	static Results ReadResults(this string[] lines)
	{
		var (linesM, linesO, linesT) = IsolateLines(lines);
		return new Results(
			linesM.Parse(Market.Parse),
			linesO.Parse(OrderUpdate.Parse),
			linesT.Parse(Trade.Parse)
		);
	}
	static T[] Parse<T>(this string[] lines, Func<string[], T> fun) =>
		lines
			.Select(line => fun(line.Split(new[] { ',' }, StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries)))
			.ToArray();
	static (string[], string[], string[]) IsolateLines(string[] lines)
	{
		var idxM = lines.Index().First(e => e.Item == "# Main").Index;
		var idxO = lines.Index().First(e => e.Item == "# Orders").Index;
		var idxT = lines.Index().First(e => e.Item == "# Trades").Index;
		return (
			lines.Skip(idxM + 2).Take(idxO - idxM - 2).Clean(),
			lines.Skip(idxO + 2).Take(idxT - idxO - 2).Clean(),
			lines.Skip(idxT + 2).Clean()
		);
	}
	static string[] Clean(this IEnumerable<string> source) => source.Where(e => e != string.Empty).ToArray();



	static string Fmt(this Prices prices)
	{
		var sb = new StringBuilder();
		sb.AppendLine($"        'Open'  : {prices.Open.Fmt()},");
		sb.AppendLine($"        'Close' : {prices.Close.Fmt()},");
		if (prices.High != null)
			sb.AppendLine($"        'High'  : {prices.High.Fmt()},");
		if (prices.Low != null)
			sb.AppendLine($"        'Low'   : {prices.Low.Fmt()},");
		return sb.ToString();
	}

	static string Fmt(this Order[] orders)
	{
		var sb = new StringBuilder();
		foreach (var order in orders)
		{
			var orderDate = Consts.StartDate + TimeSpan.FromDays(order.Date);
			sb.Append($"        if self.data.datetime.date() == dt.date({orderDate.Year}, {orderDate.Month}, {orderDate.Day}): ");
			sb.Append("self.");
			sb.Append(order.Dir == OrderDir.Buy ? "buy" : "sell");
			sb.Append($"(size={order.Size}, exectype=bt.Order.{order.Type}");
			if (order.Type is OrderType.Limit or OrderType.Stop or OrderType.StopLimit)
				sb.Append($", price={order.Price}");
			if (order.Type == OrderType.StopLimit)
				sb.Append($", plimit={order.PriceLimit}");
			sb.AppendLine(")");
		}
		return sb.ToString();
	}

	static string Fmt(this double[] arr) => $"[{string.Join(", ", arr.Select(e => $"{e}"))}]";
}